examples/spells/BoundedUInt.qnt

```diff
@@ -0,0 +1,607 @@
+  pure def wrappingAdd(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, wrappingAddUnsafe)
+
+  /// Unsafe wrapping integer subtraction.
+  /// Computes `lhs - rhs`, wrapping around at the boundary of the type.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def wrappingSubUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v - rhs.v
+    val adjusted = if (res < MIN) res + (MAX + 1) else res
+    { 
+      v: adjusted % (MAX + 1), 
+      error: "" 
+    }
+  }
+
+  /// Wrapping integer subtraction.
+  /// Computes `lhs - rhs`, wrapping around at the boundary of the type.
+  pure def wrappingSub(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, wrappingSubUnsafe)
+
+  /// Unsafe wrapping integer multiplication.
+  /// Computes `lhs * rhs`, wrapping around at the boundary of the type.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def wrappingMulUnsafe(lhs: T, rhs: T): T = 
+    { 
+      v: (lhs.v * rhs.v) % (MAX + 1), 
+      error: "" 
+    }
+  
+  /// Wrapping integer multiplication.
+  /// Computes `lhs * rhs`, wrapping around at the boundary of the type.
+  pure def wrappingMul(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, wrappingMulUnsafe)
+
+  /// Wrapping integer division.
+  /// Computes `lhs / rhs`. Wrapped division on unsigned types is just normal division. 
+  /// There’s no way wrapping could ever happen. 
+  /// This operator exists, so that all operations are accounted for in the wrapping operations.
+  pure def wrappingDiv(lhs: T, rhs: T): T = checkedDiv(lhs, rhs)
+
+  /// Wrapping integer remainder.
+  /// Computes `lhs % rhs`. Wrapped remainder on unsigned types is just normal remainder. 
+  /// There’s no way wrapping could ever happen. 
+  /// This operator exists, so that all operations are accounted for in the wrapping operations.
+  pure def wrappingRem(lhs: T, rhs: T): T = checkedRem(lhs, rhs)
+
+  /// Unsafe wrapping exponentiation.
+  /// Computes `lhs ^ rhs`, wrapping around at the boundary of the type.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def wrappingPowUnsafe(lhs: T, rhs: T): T = 
+    { 
+      v: (lhs.v ^ rhs.v) % (MAX + 1), 
+      error: "" 
+    }
+  
+  /// Wrapping exponentiation.
+  /// Computes `lhs ^ rhs`, wrapping around at the boundary of the type.
+  pure def wrappingPow(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, wrappingPowUnsafe)
+}
+
+module BoundedUInt_Test {
+  import BoundedUInt.*
+
+  /////////////
+  // CHECKED //
+  /////////////
+
+  pure def MkCInv(x: int, y: int, op: (T,T) => T, expectedErr: str, expectedVal: (int, int) => int): bool = {
+    val xx = UInt(x)
+    val yy = UInt(y)
+    val res = op(xx, yy)
+    if (xx.isValid() and yy.isValid())
+      if (res.isValid()) res.v == expectedVal(xx.v, yy.v)
+      else res.error == expectedErr
+    else res.error == "out of range"
+  }
+
+  // Checked add
+  pure def CAddInv(x: int, y: int): bool = MkCInv(x,y, checkedAdd, "overflow", (xx, yy) => xx + yy)
+
+  pure val CAddInvsTest = and {
```

Comment:
I think these tests might be too generalized, in the sense that `CAddInv` (and its equivalents for other operators) re-define a lot of the operation from the real operators. I think if we actually have explicit expectations for each scenario, the tests would be more valuable both as reference material and as enforcers of behavior:

```bluespec
pure val CAddInvsTest = and {
  checkedAdd(UInt(0), UInt(0)).v == 0,
  checkedAdd(UInt(1), UInt(0)).v == 1,
  ...,
  checkedAdd(UInt(1), UInt(MAX)).error == "overflow",
  ...
}
```

Reply:
The tests, as they are now, literally check for everything in the above example, except with less duplication, _and_ with error enforcement.
For example, `checkedAdd(UInt(0), UInt(0)).v == 0` would return true on `{v: 0, error: "some error"}`

Reply:
I had the exact some thought as Gabriela here, but was too lazy to mention it :D 

I find these tests are less clear than what Gabiela suggests, and the duplication of the logic makes them less good as tests, imo. In my POV, what Gabiela suggests would provide better checks and better explanatory value.

Code reuse in tests is only helpful if it makes the tests clearer, imo, but with the current set up there is no way for a reader to know which cases are supposed to result in which outcomes except by executing the test helper in their head.

> For example, checkedAdd(UInt(0), UInt(0)).v == 0 would return true on {v: 0, error: "some error"}

The test for this should be `checkedAdd(UInt(0), UInt(0)).error == "the error", right?

Reply:
I agree with @bugarela here. It would be easier to understand non-generalized tests. If it takes longer to understand a test than to understand the implementation, then our tests are not that useful.

Reply:
This is a nice maxim:

> If it takes longer to understand a test than to understand the implementation, then our tests are not that useful